### PR TITLE
WIP Github Integration

### PR DESCRIPTION
**Pull request by hypeJunction**
- Allows importing github releases when plugin project is first created
- Allows releases to be pushed via webhooks from Github

### CODE REVIEW

**Comment by hypeJunction** [comment]
Here is the first stab. I haven't had a chance to test the webhooks integration - need an Elgg installation with community plugins installed on the server somewhere (if anyone has time to deploy it somewhere, that would be appreciated).

File: actions/plugins/create_project.php [hunk 1]
```diff
@@ -81,35 +67,42 @@
...
+if (!empty($github_releases)) {
+	foreach ($github_releases as $github_release) {
+		$release = $plugin_project->addReleaseFromGithub($github_release);
+		if ($release) {
+			$releases[] = $release;
+		}
+	}
+} else if ($_FILES['upload']['error'] == UPLOAD_ERR_OK) {
+	$release = $plugin_project->addReleaseFromUpload('upload', [
+		'release_notes' => get_input('release_notes', ''),
```
Comment (hypeJunction) [inline actions/plugins/create_project.php]: Need to wrap with `plugins_strip_tags()`

File: actions/plugins/create_project.php [hunk 2]
```diff
@@ -81,35 +67,42 @@
...
+		'elgg_version' => get_input('elgg_version', []),
+		'comments' => get_input('comments', 'yes'),
+		'version' => strip_tags(get_input('version', '0.0')),
+		'recommended' => get_input('recommended', []),
+		'access_id' => get_input('release_access_id', ACCESS_PUBLIC),
```
Comment (hypeJunction) [inline actions/plugins/create_project.php]: Default to project access id

File: actions/plugins/create_release.php [hunk 3]
```diff
@@ -84,15 +84,7 @@
 }
 
 $release->setRecommended($recommended);
-
-$release->setHash();
```
Comment (hypeJunction) [inline actions/plugins/create_release.php]: Hash is now set for all releases during `save()`. It was missing in some actions.

File: classes/PluginRelease.php [hunk 4]
```diff
@@ -93,26 +133,82 @@ public function setHash() {
...
+	 */
+	public function createRiverEntry() {
+		if (!$this->guid) {
+			return false;
 		}
+		return elgg_create_river_item(array(
+			'view' => 'river/object/plugin_release/create',
+			'action_type' => 'create',
+			'object_guid' => $this->guid,
+		));
```
Comment (hypeJunction) [inline classes/PluginRelease.php]: Set subject if not logged in (we don't have a user when digesting hooks)

File: classes/PluginProject.php [hunk 5]
```diff
@@ -358,11 +365,283 @@ static public function moveFilesOnSystem($file, $new_owner_guid) {
...
+		$payload = $api->filterPayload($this->github_secret);
+
+		$event = _elgg_services()->request->headers->get('X-GitHub-Event');
+		$delivery = _elgg_services()->request->headers->get('X-Github-Delivery');
+
+		switch ($event) {
+			case 'release' :
+				if ($payload['action'] == 'published' && isset($payload['release'])) {
+					$github_release_id = $payload['release']['id'];
+					$release = $this->addReleaseFromGithub($github_release_id, [], true);
```
Comment (hypeJunction) [inline classes/PluginProject.php]: Pass `owner_guid` explicitly via `$attrs`

File: views/default/plugins/forms/project_github_segment.php [hunk 6]
```diff
@@ -0,0 +1,79 @@
+<?php
+
+$sticky_values = elgg_get_sticky_values('community_plugins');
+
+elgg_require_js('plugins/forms/project_github_segment');
+
+$project = elgg_extract('project', $vars);
+
+if ($project) {
```
Comment (hypeJunction) [inline views/default/plugins/forms/project_github_segment.php]: Doh `!$project`

File: classes/Elgg/CommunityPlugins/GithubService.php [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+		$request = _elgg_services()->request;
+
+		$signature = $request->headers->get('X-Hub-Signature');
+		list($algo, $hash) = explode('=', $signature, 2);
+
+		$payload = file_get_contents('php://input');
+
+		$hmac = _elgg_services()->crypto->getHmac($payload, $algo, $secret);
+
+		if (!$hmac->matchesToken($hash)) {
```
Comment (hypeJunction) [inline classes/Elgg/CommunityPlugins/GithubService.php]: @mrclay This should work, right?

Comment (mrclay) [inline classes/Elgg/CommunityPlugins/GithubService.php]: Where is the `X-Hub-Signature` header specified/created?

Comment (hypeJunction) [inline classes/Elgg/CommunityPlugins/GithubService.php]: https://developer.github.com/webhooks/#payloads

Comment (hypeJunction) [inline classes/Elgg/CommunityPlugins/GithubService.php]: We will generate a unique secret for each plugin. Plugin dev will have to setup a webhook on Github providing the payload URL (/plugins/gh/$guid) and the secret.

**Comment by hypeJunction** [comment]
The working demo is at http://pragueprogrammes.com/demo/
Username is plugin_author. Password is Elgg plus three digits of this PR number.

There is currently a slight problem with releases made by scripts (things should work fine with manual releases on Github). http://stackoverflow.com/questions/37437581/listening-to-release-asset-changes-with-github-webhooks

File: classes/Elgg/CommunityPlugins/GithubService.php (hunk 7 above)
Comment (hypeJunction) [inline classes/Elgg/CommunityPlugins/GithubService.php]: It didn't work with raw hmac, so had to change a few things

**Comment by hypeJunction** [comment]
@jdalsem @jeabakker Could you take it for a spin with one of your plugins? 
I will need to update my release workflow to first create a draft, and then publish once the upload finishes.

**Pull request closed without merging**